redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,639 @@
+///
+/// The `ScanStateWrapper` struct is a wrapper for managing the state of a scan operation in a cluster.
+/// It holds a reference to the scan state and provides methods for accessing the state.
+///
+/// The `ClusterInScan` trait defines the methods for interacting with a Redis cluster during scanning,
+/// including retrieving node information, refreshing slot mapping, and routing commands to specific nodes.
+///
+/// The `ScanState` struct represents the state of a scan operation in a Redis cluster.
+/// It holds information about the current scan state, including the cursor position, scanned slots map,
+/// slots in the scan map, node being scanned, and node's epoch.
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use crate::aio::ConnectionLike;
+use crate::cluster_async::{ClusterConnInner, Connect, InnerCore};
+use crate::cluster_routing::SlotAddr;
+use crate::{cmd, from_redis_value, Cmd, ErrorKind, RedisError, RedisResult, Value};
+
+const BITS_PER_U64: usize = 64;
+const NUM_OF_SLOTS: usize = 16384;
+const BITS_ARRAY_SIZE: usize = NUM_OF_SLOTS / BITS_PER_U64;
+const END_OF_SCAN: u16 = NUM_OF_SLOTS as u16 + 1;
+type SlotsBitsArray = [u64; BITS_ARRAY_SIZE];
+
+#[derive(PartialEq, Debug, Clone)]
+pub struct ScanState {
+    pub cursor: u64,
+    pub scanned_slots_map: SlotsBitsArray,
+    pub slots_in_scan_map: SlotsBitsArray,
+    pub node_in_scan: String,
+    pub node_epoch: u64,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ClusterScanArgs {
+    scan_state_ref: ScanStateWrapper,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<String>,
+}
+
+impl ClusterScanArgs {
+    pub(crate) fn new(
+        scan_state_ref: ScanStateWrapper,
+        match_pattern: Option<String>,
+        count: Option<usize>,
+        object_type: Option<String>,
+    ) -> Self {
+        Self {
+            scan_state_ref,
+            match_pattern,
+            count,
+            object_type,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// A wrapper struct for managing the state of a scan operation in a cluster.
+pub struct ScanStateWrapper {
+    scan_atate_ref: Arc<Option<ScanState>>,
+}
+
+impl PartialEq for ScanStateWrapper {
+    fn eq(&self, other: &Self) -> bool {
+        self.is_none() == other.is_none()
+    }
+}
+
+impl Default for ScanStateWrapper {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ScanStateWrapper {
+    fn from(scan_state: &ScanState) -> Self {
+        Self {
+            scan_atate_ref: Arc::new(Some(scan_state.clone())),
+        }
+    }
+
+    /// Creates a new instance of `ScanStateWrapper`.
+    ///
+    /// This method initializes the `ScanStateWrapper` with a read-write lock
+    /// that is initially set to `None`.
+    pub fn new() -> Self {
+        Self {
+            scan_atate_ref: Arc::new(None),
+        }
+    }
+
+    /// Returns `true` if the scan state is empty, indicating that the scan operation has finished or has not started.
```

Comment:
I don't think it is very informative to the user, when should the user use it?
maybe change it to get_state and have an enum with three values - NotStarted, InProgress and Finished.

Reply:
having a state enum will allow the user have a clear code to check when the scan is ended

Reply:
Hmm the issue is that its differ from scan method.
And basically notstarted and finished are the same. also for us to return the value its problematic because both are the same, just empty ref.
I can maybe go for inProgress and emptyProgress or something similar that fits better the actual impl.